### PR TITLE
Profile photo fetcher

### PR DESCRIPTION
**Pull request by elfametesar**

### CODE REVIEW

**Comment by elfametesar** [comment]
omg it actually passed lol

File: getpfp.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    def __init__(self):
+        self.name = _("Profile Photo Fetcher")
+
+    async def client_ready(self, client, db):
+        self.client = client
+        self.db = db
+
+    async def getpfpcmd(self, message):
+        """Fetches the profile photos of replied users, channels or chats"""
+        id = utils.get_args_raw(message)
```
Comment (penn5) [inline getpfp.py:41]: Don't override builtins

File: filters.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        self._db = db
+        self._client = client
+        self._me = await client.get_me()
+
+    async def filtercmd(self, message):
+        """Adds a filter into the list."""
+        args = (utils.get_args_raw(message)).strip()
+        reply = await message.get_reply_message()
+        chatid = str(message.chat_id)
+        filters = self._db.get("FilterModule", "filters")
```
Comment (penn5) [inline filters.py:47]: Use __name__ ffs

File: filters.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        if not args:
+            await message.edit("<b>Enter a name for the filter first!</b>")
```
Comment (penn5) [inline filters.py:49]: Make the translatable

File: filters.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            return
+        else:
+            args = args.split(" ")
```
Comment (penn5) [inline filters.py:52]: Use get_args

File: filters.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            name = args[0]
+        if not filters:
+            filters = {}
```
Comment (penn5) [inline filters.py:55]: Remove this after setting a default value

File: filters.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            filters.setdefault(chatid, {})
```
Comment (penn5) [inline filters.py:56]: Don't make this conditional

File: filters.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        if chatid not in filters:
+            filters.setdefault(chatid, {})
+        if not reply:
+            if len(args) == 1:
```
Comment (penn5) [inline filters.py:60]: Args should be from get_args_raw so it should say `if not args`

File: filters.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                await message.edit("<b>Please reply to a message or enter a text to save as filter.!</b>")
```
Comment (penn5) [inline filters.py:61]: Remove the !

File: filters.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                return
+            else:
+                value = " ".join(args[1::])
```
Comment (penn5) [inline filters.py:64]: Well in that case you should be using get_args

File: filters.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+                msg_to_log = await self._db.store_asset(value)
```
Comment (penn5) [inline filters.py:65]: No no no!!!! You should save a Message, got by get_reply_message

File: filters.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        else:
+            value = reply.text
```
Comment (penn5) [inline filters.py:67]: ???

File: filters.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            msg_to_log = await self._db.store_asset(reply)
+        filters[chatid][name] = str(msg_to_log)
```
Comment (penn5) [inline filters.py:69]: You can store ints in the db ffs

File: filters.py [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+        self._db.set("FilterModule", "filters", filters)
```
Comment (penn5) [inline filters.py:70]: `__name__`

File: filters.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        await message.edit("<b>Text ''{}'' has been successfully filtered in this chat.</b>".format(name))
```
Comment (penn5) [inline filters.py:71]: N O. just write "filter added"

Comment (penn5) [reply filters.py:71]: And remember to translate it

File: filters.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+
+    async def stopcmd(self, message):
+        """Removes a filter from the list."""
+        filtern = str(utils.get_args_raw(message))
+        filters = self._db.get("FilterModule", "filters")
```
Comment (penn5) [inline filters.py:76]: __name__

File: filters.py [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+        chatid = str(message.chat_id)
```
Comment (penn5) [inline filters.py:77]: Use ints

File: filters.py [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+        else:
+            value = reply.text
+            msg_to_log = await self._db.store_asset(reply)
+        filters[chatid][name] = str(msg_to_log)
+        self._db.set("FilterModule", "filters", filters)
+        await message.edit("<b>Text ''{}'' has been successfully filtered in this chat.</b>".format(name))
+
+    async def stopcmd(self, message):
+        """Removes a filter from the list."""
+        filtern = str(utils.get_args_raw(message))
```
Comment (penn5) [inline filters.py:75]: Don't str() a str

File: filters.py [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        filters = self._db.get("FilterModule", "filters")
+        chatid = str(message.chat_id)
+        if not filtern:
+            await message.edit("<b>Please specify the name of the filter.</b>")
+            return
+        try:
+            del filters[chatid][filtern]
+            await message.edit("<b>Filter ''{}'' successfully removed from the chat.</b>".format(filtern))
```
Comment (penn5) [inline filters.py:83]: Put this at the end of the function

File: filters.py [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+            self._db.set("FilterModule", "filters", filters)
```
Comment (penn5) [inline filters.py:84]: __name__

File: filters.py [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+        except Exception:
```
Comment (penn5) [inline filters.py:85]: NO NO NO NO NO NEVER EVER DO THIS

1. You should be only catching the line that throws, i.e. `del filters[...`
2. You should use an `else` clause on the if to run code only if it worked
3. Remove the formatting from the strings
4. Make the strings translatable
5. NEVER `EXCEPT EXCEPTION`. You want `except KeyError`

File: filters.py [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+            await message.edit("<b>Filter ''{}'' not found in this chat</b>".format(filtern))
+
+    async def stopallcmd(self, message):
+        """Clears out the filter list."""
+        filters = self._db.get("FilterModule", "filters")
```
Comment (penn5) [inline filters.py:90]: __name__

File: filters.py [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+        chatid = str(message.chat_id)
+        try:
+            del filters[chatid]
+            self._db.set("FilterModule", "filters", filters)
+            await message.edit("<b>All filters successfully removed from the chat.</b>")
+        except Exception:
```
Comment (penn5) [inline filters.py:96]: KeyError and all the shit above

File: filters.py [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+            await message.edit("<b>Filter ''{}'' not found in this chat</b>".format(filtern))
+
+    async def stopallcmd(self, message):
+        """Clears out the filter list."""
+        filters = self._db.get("FilterModule", "filters")
+        chatid = str(message.chat_id)
+        try:
+            del filters[chatid]
+            self._db.set("FilterModule", "filters", filters)
```
Comment (penn5) [inline filters.py:94]: __name__

File: filters.py [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+            del filters[chatid][filtern]
+            await message.edit("<b>Filter ''{}'' successfully removed from the chat.</b>".format(filtern))
+            self._db.set("FilterModule", "filters", filters)
+        except Exception:
+            await message.edit("<b>Filter ''{}'' not found in this chat</b>".format(filtern))
+
+    async def stopallcmd(self, message):
+        """Clears out the filter list."""
+        filters = self._db.get("FilterModule", "filters")
+        chatid = str(message.chat_id)
```
Comment (penn5) [inline filters.py:91]: Ints

File: filters.py [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+            del filters[chatid]
+            self._db.set("FilterModule", "filters", filters)
+            await message.edit("<b>All filters successfully removed from the chat.</b>")
+        except Exception:
+            await message.edit("<b>There are no filters to clear out in this chat.</b>")
+
+    async def filterscmd(self, message):
+        """Shows saved filters."""
+        filters = ""
+        filt = self._db.get("FilterModule", "filters")
```
Comment (penn5) [inline filters.py:102]: __name__

File: filters.py [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+        chatid = str(message.chat_id)
```
Comment (penn5) [inline filters.py:103]: Int

File: filters.py [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+        try:
+            for i in filt[chatid]:
+                filters += "<b> -  " + str(i) + "</b>\n"
+                pass
+        except Exception:
+            pass
+        filterl = "<b>Text that you filtered in this chat: </b>\n\n" + filters
```
Comment (penn5) [inline filters.py:110]: Translateability

File: filters.py [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+        """Shows saved filters."""
+        filters = ""
+        filt = self._db.get("FilterModule", "filters")
+        chatid = str(message.chat_id)
+        try:
+            for i in filt[chatid]:
+                filters += "<b> -  " + str(i) + "</b>\n"
+                pass
+        except Exception:
+            pass
```
Comment (penn5) [inline filters.py:109]: No!!!!!!!!!!!!!!!!!!!! Just make the exception impossible, by setting a default value

File: filters.py [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+        filterl = "<b>Text that you filtered in this chat: </b>\n\n" + filters
+        if filters:
+            await message.edit(filterl)
+        else:
+            await message.edit("<b>No filters found in this chat.</b>")
+
+    async def watcher(self, message):
+        # This is to execute cmds
+        exec = True
```
Comment (penn5) [inline filters.py:118]: Don't overwrite builtins.

File: filters.py [hunk 30]
```diff
@@ -0,0 +1,154 @@
...
+        for load in self.allloaders:
+            if load.client is self._client:
```
Comment (penn5) [inline filters.py:120]: WHAT THE FUCK ARE YOU DOING!?!?!?!?!?!??!??!?!?!!??!??

File: filters.py [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+                break
+        args = str(message.text.lower())
```
Comment (penn5) [inline filters.py:122]: OH MY GOD JUST USE GET_ARGS_RAW

File: filters.py [hunk 32]
```diff
@@ -0,0 +1,154 @@
...
+        filters = self._db.get("FilterModule", "filters")
```
Comment (penn5) [inline filters.py:123]: __name__

File: filters.py [hunk 33]
```diff
@@ -0,0 +1,154 @@
...
+        chatid = str(message.chat_id)
```
Comment (penn5) [inline filters.py:124]: Int

File: filters.py [hunk 34]
```diff
@@ -0,0 +1,154 @@
...
+        try:
+            current_prefix = self._db.get("friendly-telegram.main", "command_prefix")
```
Comment (penn5) [inline filters.py:126]: WHAT THE FUCK ARE YOU UP TO?!?!?

File: filters.py [hunk 35]
```diff
@@ -0,0 +1,154 @@
...
+        except Exception:
```
Comment (penn5) [inline filters.py:127]: IT WONT GIVE A FUCKING EXCEPTION IT WILL GIVE NONE

File: filters.py [hunk 36]
```diff
@@ -0,0 +1,154 @@
...
+            current_prefix = "."
+        if filters and chatid in filters:
+            for key in filters[chatid]:
+                if key.lower() in args.lower():
+                    id = filters[chatid][key]
```
Comment (penn5) [inline filters.py:132]: Don't override builtins

File: filters.py [hunk 37]
```diff
@@ -0,0 +1,154 @@
...
+                    value = await self._db.fetch_asset(int(id))
```
Comment (penn5) [inline filters.py:133]: Remove int()

File: filters.py [hunk 38]
```diff
@@ -0,0 +1,154 @@
...
+            if load.client is self._client:
+                break
+        args = str(message.text.lower())
+        filters = self._db.get("FilterModule", "filters")
+        chatid = str(message.chat_id)
+        try:
+            current_prefix = self._db.get("friendly-telegram.main", "command_prefix")
+        except Exception:
+            current_prefix = "."
+        if filters and chatid in filters:
```
Comment (penn5) [inline filters.py:129]: Just make filters default to {} and remove the first half

File: filters.py [hunk 39]
```diff
@@ -0,0 +1,154 @@
...
+            for key in filters[chatid]:
+                if key.lower() in args.lower():
```
Comment (penn5) [inline filters.py:131]: It might be better to lower() them at save time, to optimise it and avoid lower()ing in a loop

File: filters.py [hunk 40]
```diff
@@ -0,0 +1,154 @@
...
+    async def client_ready(self, client, db):
+        self._db = db
+        self._client = client
+        self._me = await client.get_me()
+
+    async def filtercmd(self, message):
+        """Adds a filter into the list."""
+        args = (utils.get_args_raw(message)).strip()
+        reply = await message.get_reply_message()
+        chatid = str(message.chat_id)
```
Comment (penn5) [inline filters.py:46]: don't make an int into a str

File: filters.py [hunk 41]
```diff
@@ -0,0 +1,154 @@
...
+        self._ratelimit = []
+
+    async def client_ready(self, client, db):
+        self._db = db
+        self._client = client
+        self._me = await client.get_me()
+
+    async def filtercmd(self, message):
+        """Adds a filter into the list."""
+        args = (utils.get_args_raw(message)).strip()
```
Comment (penn5) [inline filters.py:44]: its already stripped for you

File: filters.py [hunk 42]
```diff
@@ -0,0 +1,154 @@
...
+                            exec = False
+                            cmd = value.message
+                        if not exec:
+                            await message.client.send_message(int(chatid), value.message, reply_to=message.id)
+                            return
+                        else:
+                            cmdspl = cmd.lower().split(" ")
+                            respond = await message.client.send_message(int(chatid), cmd, reply_to=message.id)
+                            try:
+                                await load.dispatch(cmdspl[0], respond)
```
Comment (penn5) [inline filters.py:149]: WHAT THE FUCK ARE YOU CRAZY!?!?!?!?!?!!?!?!?!?!?! DONT SHIP THIS MODULE ANYWHERE!!!!!!!!!!!!!!!!!!!!!!!!!!!!

File: filters.py [hunk 43]
```diff
@@ -0,0 +1,154 @@
...
+                            except Exception:
+                                pass
```
Comment (penn5) [inline filters.py:151]: Never catch-all
Never pass

Comment (penn5) [reply filters.py:151]: you should a) catch the right error and b) logger.exception()

File: filters.py [hunk 44]
```diff
@@ -0,0 +1,154 @@
...
+                    else:
+                        await self._client.send_message(int(chatid), value.text, file=value.media, reply_to=message.id)
```
Comment (penn5) [inline filters.py:153]: also fucked up. just do `utils.answer(message, value)`

File: getpfp.py [hunk 45]
```diff
@@ -0,0 +1,75 @@
...
+
+import logging
+
+from .. import loader, utils
+
+logger = logging.getLogger(__name__)
+
+
+def register(cb):
+    cb(GetPPMod())
```
Comment (penn5) [inline getpfp.py:27]: Can you change that to GetPFPMod? I feel like someone left a joke in production.

File: getpfp.py [hunk 46]
```diff
@@ -0,0 +1,75 @@
...
+
+    async def client_ready(self, client, db):
+        self.client = client
+        self.db = db
+
+    async def getpfpcmd(self, message):
+        """Fetches the profile photos of replied users, channels or chats"""
+        id = utils.get_args_raw(message)
+        user = await message.get_reply_message()
+        chat = message.input_chat
```
Comment (penn5) [inline getpfp.py:43]: You don't use this variable properly. It should be `(await message.get_reply_message()).input_sender if message.is_reply else message.input_chat`, then you can remove all the later conditionals

File: getpfp.py [hunk 47]
```diff
@@ -0,0 +1,75 @@
...
+        if user:  # If replied to a message, photos should be extracted from the user
+            photos = await self.client.get_profile_photos(user.sender)
+        else:
+            photos = await self.client.get_profile_photos(chat)
```
Comment (penn5) [inline getpfp.py:47]: Make this not conditional

Comment (penn5) [inline getpfp.py:47]: And make this iter_profile_photos not get_profile_photos

File: getpfp.py [hunk 48]
```diff
@@ -0,0 +1,75 @@
...
+            if len(photos) > 0:
+                await self.client.send_file(message.chat_id, photos)
+            else:
+                try:  # I had to do this bit because get_profile_photos function wasnt working for one photo
+                    if user:
+                        photo = await self.client.download_profile_photo(user.sender)
+                    else:
+                        photo = await self.client.download_profile_photo(message.input_chat)
+                    await self.client.send_file(message.chat_id, photo)
+                except Exception:
```
Comment (penn5) [inline getpfp.py:58]: OMFG

File: getpfp.py [hunk 49]
```diff
@@ -0,0 +1,75 @@
...
+                    await message.edit("<b>This user has no photos.</b>")
```
Comment (penn5) [inline getpfp.py:59]: utils.answer

File: getpfp.py [hunk 50]
```diff
@@ -0,0 +1,75 @@
...
+        id = utils.get_args_raw(message)
+        user = await message.get_reply_message()
+        chat = message.input_chat
+        if user:  # If replied to a message, photos should be extracted from the user
+            photos = await self.client.get_profile_photos(user.sender)
+        else:
+            photos = await self.client.get_profile_photos(chat)
+        if not id:  # if photo id isn't stated, it will fetch all the photos of target
+            if len(photos) > 0:
+                await self.client.send_file(message.chat_id, photos)
```
Comment (penn5) [inline getpfp.py:50]: you mean `if photos`

File: getpfp.py [hunk 51]
```diff
@@ -0,0 +1,75 @@
...
+                try:  # I had to do this bit because get_profile_photos function wasnt working for one photo
+                    if user:
+                        photo = await self.client.download_profile_photo(user.sender)
+                    else:
+                        photo = await self.client.download_profile_photo(message.input_chat)
+                    await self.client.send_file(message.chat_id, photo)
+                except Exception:
+                    await message.edit("<b>This user has no photos.</b>")
+                    return
+        else:  # If photo id is stated
```
Comment (penn5) [inline getpfp.py:61]: If you only want one, why does the above code download all of them? Take, for example, t.me/spechide. There's almost 200000 profile photos. Imagine downloading all of those, just to access one of them.

File: getpfp.py [hunk 52]
```diff
@@ -0,0 +1,75 @@
...
+            try:
+                id = int(id)
+                if id <= 0:
+                    await message.edit("<b>ID number you entered is invalid</b>")
+                    return
+            except Exception:
```
Comment (penn5) [inline getpfp.py:67]: ValueError

File: getpfp.py [hunk 53]
```diff
@@ -0,0 +1,75 @@
...
+                        photo = await self.client.download_profile_photo(user.sender)
+                    else:
+                        photo = await self.client.download_profile_photo(message.input_chat)
+                    await self.client.send_file(message.chat_id, photo)
+                except Exception:
+                    await message.edit("<b>This user has no photos.</b>")
+                    return
+        else:  # If photo id is stated
+            try:
+                id = int(id)
```
Comment (penn5) [inline getpfp.py:63]: don't replace builtins

File: getpfp.py [hunk 54]
```diff
@@ -0,0 +1,75 @@
...
+                if id <= 0:
+                    await message.edit("<b>ID number you entered is invalid</b>")
```
Comment (penn5) [inline getpfp.py:65]: utils.answer

File: getpfp.py [hunk 55]
```diff
@@ -0,0 +1,75 @@
...
+                    return
+            except Exception:
+                await message.edit("<b>ID number you entered is invalid</b>")
```
Comment (penn5) [inline getpfp.py:68]: utils.answer

File: getpfp.py [hunk 56]
```diff
@@ -0,0 +1,75 @@
...
+                return
+            if id <= len(photos):
+                send_photos = await self.client.download_media(photos[id - 1])
+                await self.client.send_file(message.chat_id, send_photos)
+            else:
+                await message.edit("<b>No photo found with that id</b>")
```
Comment (penn5) [inline getpfp.py:74]: utils.answer

File: getpfp.py [hunk 57]
```diff
@@ -0,0 +1,75 @@
...
+            try:
+                id = int(id)
+                if id <= 0:
+                    await message.edit("<b>ID number you entered is invalid</b>")
+                    return
+            except Exception:
+                await message.edit("<b>ID number you entered is invalid</b>")
+                return
+            if id <= len(photos):
+                send_photos = await self.client.download_media(photos[id - 1])
```
Comment (penn5) [inline getpfp.py:71]: This array index notation should not really be used. Instead, set `limit` and `offset` on iter_profile_photos (tbh, you should use get_profile_photos when you are fetching only 1, but without limit and offset you certainly must use iter)

File: getpfp.py [hunk 58]
```diff
@@ -0,0 +1,75 @@
...
+                await self.client.send_file(message.chat_id, send_photos)
```
Comment (penn5) [inline getpfp.py:72]: utils.answer supports this iirc

File: getpfp.py [hunk 59]
```diff
@@ -0,0 +1,75 @@
...
+            else:
+                await message.edit("<b>No photo found with that id</b>")
+                return
```
Comment (penn5) [inline getpfp.py:75]: seriously? this is the last line in the entire file.

**Review by penn5 — CHANGES_REQUESTED** [review]
Decent effort but wtf was that about the loaders

File: getpfp.py (hunk 1 above)
Comment (elfametesar) [reply getpfp.py:41]: what builtin ? if you mean the cmd you said okay to it

File: filters.py (hunk 2 above)
Comment (elfametesar) [reply filters.py:47]: what.. i didnt know i did a pr on filters too. name should stay that way btw

File: filters.py (hunk 3 above)
Comment (elfametesar) [reply filters.py:49]: alright

File: getpfp.py (hunk 59 above)
Comment (elfametesar) [reply getpfp.py:75]: oof stop making things a problem

File: filters.py (hunk 10 above)
Comment (elfametesar) [reply filters.py:65]: NO WAY, thats not the only way i get a filter value

File: filters.py (hunk 12 above)
Comment (elfametesar) [reply filters.py:69]: its been a bad habit to me from an old project

File: filters.py (hunk 17 above)
Comment (elfametesar) [reply filters.py:75]: this is funny lmao. i didnt check some parts cuz they were already working, this is my noober times

File: filters.py (hunk 18 above)
Comment (elfametesar) [reply filters.py:83]: makes sense, i should have thought of that

File: getpfp.py (hunk 1 above)
Comment (penn5) [reply getpfp.py:41]: The name `id` is taken by python

File: filters.py (hunk 2 above)
Comment (penn5) [reply filters.py:47]: The name must change.

File: filters.py (hunk 10 above)
Comment (penn5) [reply filters.py:65]: Message objects are mutable. Edit the text to the target (not with .edit(), but with .message = ...)

File: filters.py (hunk 12 above)
Comment (penn5) [reply filters.py:69]: Ok, get rid of it

File: filters.py (hunk 29 above)
Comment (penn5) [reply filters.py:118]: Exec is also a reserved word, like id

File: getpfp.py (hunk 59 above)
Comment (penn5) [reply getpfp.py:75]: It's a redundant line, delete it

**Review by penn5 — COMMENTED** [review]
Can't review this much tonight. Glancing, looks good but it might be shit, idk

**Comment by penn5** [comment]
There's a conflict that must be resolved. Rebase your changes

**Pull request closed without merging**